File: apps/api/src/auth/jwt.rs
use jsonwebtoken::{decode, decode_header, Algorithm, DecodingKey, Validation};
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Claims {
    pub sub: String,
    pub email: String,
    pub exp: usize,
    pub iat: usize,
    pub aud: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
struct Jwk {
    kid: String,
    #[allow(dead_code)]
    kty: String,
    #[serde(rename = "use")]
    #[allow(dead_code)]
    use_: Option<String>,
    n: Option<String>,
    e: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
struct Jwks {
    keys: Vec<Jwk>,
}

pub struct JwtValidator {
    team_domain: String,
    aud: String,
    hmac_secret: String,
    cache: RwLock<Option<(Jwks, std::time::Instant)>>,
    client: reqwest::Client,
}

const JWKS_CACHE_TTL: std::time::Duration = std::time::Duration::from_secs(6 * 60 * 60);

impl JwtValidator {
    pub fn new(team_domain: String, aud: String) -> Self {
        Self {
            team_domain,
            aud,
            hmac_secret: std::env::var("JWT_SECRET").unwrap_or_else(|_| "fitmentor-dev-secret-change-in-production".to_string()),
            cache: RwLock::new(None),
            client: reqwest::Client::new(),
        }
    }

    async fn fetch_jwks(&self) -> Result<Jwks, String> {
        let url = format!(
            "https://{}/cdn-cgi/access/certs",
            self.team_domain
        );
        let resp = self
            .client
            .get(&url)
            .send()
            .await
            .map_err(|e| format!("failed to fetch JWKS: {e}"))?;
        let jwks: Jwks = resp
            .json()
            .await
            .map_err(|e| format!("failed to parse JWKS: {e}"))?;
        Ok(jwks)
    }

    async fn get_jwks(&self) -> Result<Jwks, String> {
        {
            let cache = self.cache.read().await;
            if let Some((jwks, ts)) = cache.as_ref() {
                if ts.elapsed() < JWKS_CACHE_TTL {
                    return Ok(jwks.clone());
                }
            }
        }
        let jwks = self.fetch_jwks().await?;
        {
            let mut cache = self.cache.write().await;
            *cache = Some((jwks.clone(), std::time::Instant::now()));
        }
        Ok(jwks)
    }

    fn decode_jwt_header_kid(token: &str) -> Result<String, String> {
        let header = decode_header(token)
            .map_err(|e| format!("failed to decode JWT header: {e}"))?;
        header
            .kid
            .ok_or_else(|| "JWT header missing kid".to_string())
    }

    fn find_jwk<'a>(jwks: &'a Jwks, kid: &str) -> Result<&'a Jwk, String> {
        jwks.keys
            .iter()
            .find(|k| k.kid == kid)
            .ok_or_else(|| format!("no matching JWK for kid: {kid}"))
    }

    fn build_decoding_key(jwk: &Jwk) -> Result<DecodingKey, String> {
        let n = jwk.n.as_ref().ok_or("JWK missing n")?;
        let e = jwk.e.as_ref().ok_or("JWK missing e")?;
        DecodingKey::from_rsa_components(n, e).map_err(|e| format!("invalid RSA key: {e}"))
    }

    async fn validate_cloudflare_access(&self, token: &str) -> Result<Claims, String> {
        let kid = Self::decode_jwt_header_kid(token)?;
        let jwks = self.get_jwks().await?;
        let jwk = Self::find_jwk(&jwks, &kid)?;
        let key = Self::build_decoding_key(jwk)?;

        let mut validation = Validation::new(Algorithm::RS256);
        validation.set_audience(&[&self.aud]);
        validation.set_issuer(&[format!(
            "https://{}.cloudflareaccess.com",
            self.team_domain
        )]);

        let token_data = decode::<Claims>(token, &key, &validation)
            .map_err(|e| format!("JWT validation failed: {e}"))?;

        Ok(token_data.claims)
    }

    fn validate_hmac(&self, token: &str) -> Result<Claims, String> {
        let key = DecodingKey::from_secret(self.hmac_secret.as_bytes());
        let mut validation = Validation::new(Algorithm::HS256);
        validation.validate_exp = true;

        let token_data = decode::<Claims>(token, &key, &validation)
            .map_err(|e| format!("HMAC JWT validation failed: {e}"))?;

        Ok(token_data.claims)
    }

    pub async fn validate(&self, token: &str) -> Result<Claims, String> {
        if let Ok(claims) = self.validate_hmac(token) {
            return Ok(claims);
        }
        self.validate_cloudflare_access(token).await
    }
}

File: api/internal/auth/jwt/jwt.go
// Package jwt implements the JWT validation from auth/jwt.rs:
// HMAC first, then Cloudflare Access RS256 via JWKS.
package jwt

import (
	"crypto"
	"crypto/hmac"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"strings"
	"sync"
	"time"
)

type Claims struct {
	Sub   string `json:"sub"`
	Email string `json:"email"`
	Exp   int64  `json:"exp"`
	Iat   int64  `json:"iat"`
	Aud   any    `json:"aud"`
}

type jwk struct {
	Kid string  `json:"kid"`
	N   *string `json:"n"`
	E   *string `json:"e"`
}

type jwks struct {
	Keys []jwk `json:"keys"`
}

const jwksCacheTTL = 6 * time.Hour

// hmacLeeway mirrors jsonwebtoken's default 60s leeway.
const hmacLeeway = 60

type JWTValidator struct {
	teamDomain string
	aud        string
	hmacSecret string
	httpClient *http.Client
	mu         sync.RWMutex
	cachedJWKS *jwks
	cachedAt   time.Time
}

func NewJWTValidator(teamDomain, aud, hmacSecret string) *JWTValidator {
	return &JWTValidator{
		teamDomain: teamDomain,
		aud:        aud,
		hmacSecret: hmacSecret,
		httpClient: &http.Client{Timeout: 10 * time.Second},
	}
}

func b64urlDecode(s string) ([]byte, error) {
	if m := len(s) % 4; m != 0 {
		s += strings.Repeat("=", 4-m)
	}
	return base64.URLEncoding.DecodeString(s)
}

func splitToken(token string) (headerB64, payloadB64, sigB64 string, ok bool) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return "", "", "", false
	}
	return parts[0], parts[1], parts[2], true
}

func decodeHeader(token string) (kid, alg string, err error) {
	hB64, _, _, ok := splitToken(token)
	if !ok {
		return "", "", fmt.Errorf("malformed token")
	}
	raw, err := b64urlDecode(hB64)
	if err != nil {
		return "", "", fmt.Errorf("bad header: %w", err)
	}
	var h struct {
		Kid string `json:"kid"`
		Alg string `json:"alg"`
	}
	if err := json.Unmarshal(raw, &h); err != nil {
		return "", "", err
	}
	return h.Kid, h.Alg, nil
}

func decodePayload(token string) (Claims, error) {
	_, pB64, _, ok := splitToken(token)
	if !ok {
		return Claims{}, fmt.Errorf("malformed token")
	}
	raw, err := b64urlDecode(pB64)
	if err != nil {
		return Claims{}, err
	}
	var c Claims
	if err := json.Unmarshal(raw, &c); err != nil {
		return Claims{}, err
	}
	return c, nil
}

func audContains(aud any, want string) bool {
	switch v := aud.(type) {
	case string:
		return v == want
	case []any:
		for _, e := range v {
			if s, ok := e.(string); ok && s == want {
				return true
			}
		}
	}
	return false
}

func (v *JWTValidator) validateHMAC(token string) (Claims, error) {
	hB64, pB64, sB64, ok := splitToken(token)
	if !ok {
		return Claims{}, fmt.Errorf("malformed token")
	}
	if _, alg, err := decodeHeader(token); err != nil || alg != "HS256" {
		return Claims{}, fmt.Errorf("not HS256")
	}
	mac := hmac.New(sha256.New, []byte(v.hmacSecret))
	mac.Write([]byte(hB64 + "." + pB64))
	sig, err := b64urlDecode(sB64)
	if err != nil {
		return Claims{}, err
	}
	if !hmac.Equal(mac.Sum(nil), sig) {
		return Claims{}, fmt.Errorf("bad signature")
	}
	c, err := decodePayload(token)
	if err != nil {
		return Claims{}, err
	}
	if c.Exp == 0 || time.Now().Unix() > c.Exp+hmacLeeway {
		return Claims{}, fmt.Errorf("expired")
	}
	return c, nil
}

func (v *JWTValidator) getJWKS() (*jwks, error) {
	v.mu.RLock()
	if v.cachedJWKS != nil && time.Since(v.cachedAt) < jwksCacheTTL {
		defer v.mu.RUnlock()
		return v.cachedJWKS, nil
	}
	v.mu.RUnlock()

	resp, err := v.httpClient.Get("https://" + v.teamDomain + "/cdn-cgi/access/certs")
	if err != nil {
		return nil, fmt.Errorf("failed to fetch JWKS: %w", err)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var parsed jwks
	if err := json.Unmarshal(body, &parsed); err != nil {
		return nil, fmt.Errorf("failed to parse JWKS: %w", err)
	}
	v.mu.Lock()
	v.cachedJWKS = &parsed
	v.cachedAt = time.Now()
	v.mu.Unlock()
	return &parsed, nil
}

func rsaKeyFromJWK(k jwk) (*rsa.PublicKey, error) {
	if k.N == nil || k.E == nil {
		return nil, fmt.Errorf("JWK missing n/e")
	}
	nBytes, err := b64urlDecode(*k.N)
	if err != nil {
		return nil, err
	}
	eBytes, err := b64urlDecode(*k.E)
	if err != nil {
		return nil, err
	}
	n := new(big.Int).SetBytes(nBytes)
	var e int
	if len(eBytes) < 8 {
		padded := make([]byte, 8)
		copy(padded[8-len(eBytes):], eBytes)
		e = int(int64(binary.BigEndian.Uint64(padded)))
	} else {
		e = int(new(big.Int).SetBytes(eBytes).Int64())
	}
	return &rsa.PublicKey{N: n, E: e}, nil
}

func (v *JWTValidator) validateCloudflare(token string) (Claims, error) {
	kid, alg, err := decodeHeader(token)
	if err != nil {
		return Claims{}, err
	}
	if kid == "" {
		return Claims{}, fmt.Errorf("JWT header missing kid")
	}
	if alg != "RS256" {
		return Claims{}, fmt.Errorf("unexpected alg %q", alg)
	}
	set, err := v.getJWKS()
	if err != nil {
		return Claims{}, err
	}
	var match *jwk
	for i := range set.Keys {
		if set.Keys[i].Kid == kid {
			match = &set.Keys[i]
			break
		}
	}
	if match == nil {
		return Claims{}, fmt.Errorf("no matching JWK for kid: %s", kid)
	}
	pub, err := rsaKeyFromJWK(*match)
	if err != nil {
		return Claims{}, fmt.Errorf("invalid RSA key: %w", err)
	}
	hB64, pB64, sB64, _ := splitToken(token)
	sig, err := b64urlDecode(sB64)
	if err != nil {
		return Claims{}, err
	}
	digest := sha256.Sum256([]byte(hB64 + "." + pB64))
	if err := rsa.VerifyPKCS1v15(pub, crypto.SHA256, digest[:], sig); err != nil {
		return Claims{}, fmt.Errorf("JWT validation failed: %w", err)
	}
	c, err := decodePayload(token)
	if err != nil {
		return Claims{}, err
	}
	if c.Exp == 0 || time.Now().Unix() > c.Exp+hmacLeeway {
		return Claims{}, fmt.Errorf("expired")
	}
	if !audContains(c.Aud, v.aud) {
		return Claims{}, fmt.Errorf("bad audience")
	}
	// Issuer check mirrors jsonwebtoken set_issuer(["https://{team}.cloudflareaccess.com"]).
	// The issuer lives in the payload; verify it explicitly.
	_, pRaw, _, _ := splitToken(token)
	if raw, err := b64urlDecode(pRaw); err == nil {
		var m map[string]any
		if json.Unmarshal(raw, &m) == nil {
			wantIss := "https://" + v.teamDomain + ".cloudflareaccess.com"
			if iss, _ := m["iss"].(string); iss != wantIss {
				return Claims{}, fmt.Errorf("bad issuer")
			}
		}
	}
	return c, nil
}

// Validate tries HMAC first, then Cloudflare Access (mirrors jwt.rs).
func (v *JWTValidator) Validate(token string) (Claims, error) {
	if c, err := v.validateHMAC(token); err == nil {
		return c, nil
	}
	return v.validateCloudflare(token)
}
